File: backend/app/schemas/field_layout.py
from pydantic import BaseModel


class StudyInfo(BaseModel):
    studyDbId: str
    studyName: str
    rows: int = 0
    cols: int = 0
    programDbId: str | None = None
    locationDbId: str | None = None
    startDate: str | None = None
    endDate: str | None = None

class Plot(BaseModel):
    plotNumber: str
    row: int
    col: int
    germplasmName: str | None = None
    germplasmDbId: str | None = None
    blockNumber: int | None = None
    replicate: int | None = None
    entryType: str | None = None
    observationUnitDbId: str | None = None

class FieldLayoutSummary(BaseModel):
    total_plots: int
    check_plots: int
    test_plots: int
    unique_germplasm: int
    blocks: int
    replicates: int

class FieldLayoutResponse(BaseModel):
    study: StudyInfo
    plots: list[Plot]
    summary: FieldLayoutSummary

class GermplasmRef(BaseModel):
    germplasmDbId: str
    germplasmName: str

class GermplasmListResponse(BaseModel):
    data: list[GermplasmRef]

class LayoutGenerationResponse(BaseModel):
    message: str
    study_id: str
    design: str
    blocks: int
    replicates: int
    total_plots: int
    note: str | None = None

class ExportResponse(BaseModel):
    message: str
    study_id: str
    format: str
    download_url: str | None = None
    note: str | None = None

File: frontend/src/lib/api/phenotyping/field-layout.ts
import { ApiClientCore } from "../core/client";

export interface FieldStudy {
  studyDbId: string;
  studyName: string;
  rows: number;
  cols: number;
  programDbId?: string;
  locationDbId?: string;
  startDate?: string;
  endDate?: string;
}

export interface FieldPlot {
  plotNumber: string;
  row: number;
  col: number;
  germplasmName?: string;
  germplasmDbId?: string;
  blockNumber?: number;
  replicate?: number;
  entryType?: string;
  observationUnitDbId?: string;
}

export interface FieldLayoutSummary {
  total_plots: number;
  check_plots: number;
  test_plots: number;
  unique_germplasm: number;
  blocks: number;
  replicates: number;
}

export class FieldLayoutService {
  constructor(private client: ApiClientCore) {}

  async getStudies(params?: { program_id?: string; search?: string }) {
    const searchParams = new URLSearchParams();
    if (params?.program_id) searchParams.set("program_id", params.program_id);
    if (params?.search) searchParams.set("search", params.search);
    const query = searchParams.toString();
    return this.client.get<{ data: FieldStudy[]; total: number }>(
      `/api/v2/field-layout/studies${query ? `?${query}` : ""}`
    );
  }

  async getStudy(studyId: string) {
    return this.client.get<{ data: FieldStudy }>(
      `/api/v2/field-layout/studies/${studyId}`
    );
  }

  async getLayout(studyId: string) {
    return this.client.get<{
      study: FieldStudy;
      plots: FieldPlot[];
      summary: FieldLayoutSummary;
    }>(`/api/v2/field-layout/studies/${studyId}/layout`);
  }

  async getPlots(
    studyId: string,
    params?: { block?: number; replicate?: number; entry_type?: string }
  ) {
    const searchParams = new URLSearchParams();
    if (params?.block) searchParams.set("block", params.block.toString());
    if (params?.replicate)
      searchParams.set("replicate", params.replicate.toString());
    if (params?.entry_type) searchParams.set("entry_type", params.entry_type);
    const query = searchParams.toString();
    return this.client.get<{ data: FieldPlot[]; total: number }>(
      `/api/v2/field-layout/studies/${studyId}/plots${query ? `?${query}` : ""}`
    );
  }

  async getPlot(studyId: string, plotNumber: string) {
    return this.client.get<{ data: FieldPlot }>(
      `/api/v2/field-layout/studies/${studyId}/plots/${plotNumber}`
    );
  }

  async updatePlot(
    studyId: string,
    plotNumber: string,
    data: Partial<FieldPlot>
  ) {
    return this.client.put<any>(
      `/api/v2/field-layout/studies/${studyId}/plots/${plotNumber}`,
      data
    );
  }

  async getGermplasm() {
    return this.client.get<{
      data: Array<{ germplasmDbId: string; germplasmName: string }>;
    }>("/api/v2/field-layout/germplasm");
  }

  async generateLayout(
    studyId: string,
    params: { design?: string; blocks?: number; replicates?: number }
  ) {
    const searchParams = new URLSearchParams();
    if (params.design) searchParams.set("design", params.design);
    if (params.blocks) searchParams.set("blocks", params.blocks.toString());
    if (params.replicates)
      searchParams.set("replicates", params.replicates.toString());
    return this.client.post<any>(
      `/api/v2/field-layout/studies/${studyId}/generate?${searchParams}`,
      {}
    );
  }

  async exportLayout(studyId: string, format: string = "csv") {
    return this.client.get<any>(
      `/api/v2/field-layout/export/${studyId}?format=${format}`
    );
  }
}
